### ISSUE TITLE
numpy.clip does not mantain array dtype when setting min and max to int64 limits

### ISSUE THREAD

**Issue by cttlfsh** [numpy/numpy#16609]
Numpy.clip does not seem to correctly mantain the dtype of the clipped array with some clipping min and max values.
In particular, we found out that the max (and/or min) values that can be used as clipping values is 2^62, any higher number gives a different dtype wrt to the original unclipped array.

### Reproducing code example:

```python
import numpy as np

a = np.array([1,2,3,4,5])
# returns int64
print(a.dtype)

i = - int(np.exp2(64))
j = int(np.exp2(64))
ac = np.clip(a,i,j)
# returns object
print(ac.dtype)

i = - int(np.exp2(63))
j = int(np.exp2(63))
ac = np.clip(a,i,j)
# returns float64
print(ac.dtype)

i = - int(np.exp2(62))
j = int(np.exp2(62))
ac = np.clip(a,i,j)
# returns int64 again
print(ac.dtype)
```

As stated above, the output is:
```
int64
object
float64
int64
```

### Numpy/Python version information:

__numpy.version__: 1.18.1
__sys.version__: 3.8.2 (default, Mar 10 2020, 15:23:31)

**Comment by eric-wieser** [comment]
Clip isn't special here, `np.maximum(a, j)` and `np.add(a, j)` show the same problem. The issue is that `np.asarray(j).dtype` is `uint64`, and `np.result_type(uint64, int64) == float64`.

The  `np.result_type(uint64, int64) == float64` problem comes up all over the place (@seberg, do we have canonical issue to link to here?), so we should probably close this as a duplicate.

**Comment by seberg** [comment]
Gh-5745 and gh-8851, I am pretty sure there was a recent one, although maybe that was closed after explaining the behaviour.

The second (result-type) issue, maybe actually g-10322.  I also found gh-9460 and gh-13591 the latter may be as close as we have.   I have a tracking issue or two, but I doubt I ever put this on since this is value-based casting, and while I think we all have an intention to get rid of it.
